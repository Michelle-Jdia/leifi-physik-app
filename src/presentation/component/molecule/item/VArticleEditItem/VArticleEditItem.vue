<script setup lang="ts">
import type { ClassValue } from 'clsx';
import { cn } from '@/presentation/helper/style';
import { IonReorder } from '@ionic/vue';
import VCheckbox from '@/atom/checkbox/VCheckbox/VCheckbox.vue';
import VIcon from '@/atom/icon/VIcon/VIcon.vue';

interface Props {
    id: string;
    title: string;
    subtitle: string;
    withCopy?: boolean;
    modifier?: ClassValue;
}

defineProps<Props>();

const emit = defineEmits<{
    (e: 'onCheck', { id, isChecked }: { id: string; isChecked: boolean }): void;
    (e: 'onDelete', { id }: { id: string }): void;
    (e: 'onCopy', { id }: { id: string }): void;
}>();
</script>

<template>
    <div
        :class="
            cn('article-edit-item v-flex v-items-center v-rounded v-bg-white v-p-box', modifier)
        "
        lines="none"
    >
        <v-checkbox
            modifier="v-mr-auto v-w-full"
            :value="id"
            @ion-change="(event) => emit('onCheck', { id, isChecked: event.target.checked })"
        >
            {{ title }}

            <div class="v-h5 v-text-ty-helper">{{ subtitle }}</div>
        </v-checkbox>

        <div class="v-ml-box v-flex">
            <v-icon
                v-if="withCopy"
                modifier="v-mr-box-xl v-cursor-pointer"
                src="icon/copy.svg"
                data-testid="article-edit-item-copy"
                @click="emit('onCopy', { id })"
            />

            <v-icon
                modifier="v-cursor-pointer"
                src="icon/close.svg"
                data-testid="article-edit-item-close"
                @click="emit('onDelete', { id })"
            />

            <ion-reorder slot="end" class="v-ml-box-xl">
                <v-icon src="icon/dots.svg" data-testid="article-edit-item-reorder" />
            </ion-reorder>
        </div>
    </div>
</template>
