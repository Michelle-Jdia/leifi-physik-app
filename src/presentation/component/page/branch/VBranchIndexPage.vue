<script setup lang="ts">
import type { SliderItem } from '@/molecule/slider/VTopicBoxSlider/VTopicBoxSlider.vue';
import { onIonViewWillEnter } from '@ionic/vue';
import { search } from 'ionicons/icons';
import { useBranches } from '@/store/branchStore';
import { getTopicsByBranch, createTopicSlider } from '@/store/topicStore';
import { createReactiveData } from '@/connection/helper/fetcher';
import VBranchHeader from '@/layout/header/VBranchHeader/VBranchHeader.vue';
import VPage from '@/layout/page/VPage.vue';
import VTopicBoxSlider from '@/molecule/slider/VTopicBoxSlider/VTopicBoxSlider.vue';
import VBranchBox from '@/atom/box/VBranchBox/VBranchBox.vue';
import VLoader from '@/atom/loader/VLoader/VLoader.vue';
import VInput from '@/atom/input/VInput/VInput.vue';

type TopicSliderItem = Record<string, SliderItem[]>;

const { data: branches, getData } = useBranches();

const { data: sliders, getData: buildSliders } = createReactiveData<never, TopicSliderItem | void>(
    async () => {
        await getData();

        if (!branches.value) {
            return;
        }

        const promises = branches.value.map((branch) => {
            return getTopicsByBranch({
                params: {
                    branchId: branch.id,
                },
            }).getData();
        });

        const responses = await Promise.all(promises);

        return responses.reduce((sliders: TopicSliderItem, response) => {
            if (response && response.length) {
                // @ts-ignore @todo figure out why its complining
                const branchId = response[0].referenced_branch;

                if (!sliders[branchId]) {
                    sliders[branchId] = response.map(createTopicSlider);
                }
            }

            return sliders;
        }, {});
    },
)();

onIonViewWillEnter(buildSliders);
</script>

<template>
    <v-page>
        <v-loader :logs="['useBranches', 'getTopicsByBranch']" />

        <template #header>
            <v-branch-header>Alle Inhalte</v-branch-header>
        </template>

        <v-input placeholder="Alle Inhalte durchsuchen oder Code eingeben..." :icon="search" />

        <h1 class="v-h2 v-mt-box-xl v-font-bold">Teilgebiete der Physik</h1>

        <div class="v-mt-box-xl v-grid v-grid-cols-2 v-gap-box-md md:v-grid-cols-4">
            <template v-for="branch in branches" :key="branch.name">
                <router-link :to="{ name: 'branch-id', params: { id: branch.id } }">
                    <v-branch-box
                        :img="{
                            src: branch.icon?.url || '',
                            alt: branch.icon?.alt || '',
                        }"
                    >
                        {{ branch.name }}
                    </v-branch-box>
                </router-link>
            </template>
        </div>

        <v-topic-box-slider
            v-for="branch in branches"
            :key="branch.id"
            :modifier="'v-mt-box-md v-bg-' + branch.color"
            :branch-img="{
                src: branch.icon?.url || '',
                alt: branch.icon?.alt || '',
            }"
            :link="{ name: 'branch-id', params: { id: branch.id } }"
            :slider-items="(sliders && sliders[branch.id]) || []"
        >
            {{ branch.name }}
        </v-topic-box-slider>
    </v-page>
</template>
