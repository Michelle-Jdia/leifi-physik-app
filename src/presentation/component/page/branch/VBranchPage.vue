<script setup lang="ts">
import type { ComputedRef } from 'vue';
import type { SliderItem } from '@/molecule/slider/VTopicBoxSlider/VTopicBoxSlider.vue';
import { computed } from 'vue';
import { useBranch } from '@/store/branchStore';
import {
    TopicDegree,
    getTopicsByBranch,
    createTopicSlider,
    isTopicFirstDegree,
    isTopicSecondDegree,
    isTopicThirdDegree,
} from '@/store/topicStore';
import VPage from '@/layout/page/VPage.vue';
import VBranchHeader from '@/layout/header/VBranchHeader/VBranchHeader.vue';
import VTopicBoxSlider from '@/molecule/slider/VTopicBoxSlider/VTopicBoxSlider.vue';
import VLoader from '@/atom/loader/VLoader/VLoader.vue';

interface Props {
    id: string;
}

const props = defineProps<Props>();

type TopicSliderItem = Record<string, SliderItem[]>;

const themeTitle = 'Themenbereiche';

const { data: branch, getData: getBranch } = useBranch({
    params: {
        id: props.id,
    },
});

const { data: topics, getData: getTopics } = getTopicsByBranch({
    params: {
        branchId: props.id,
    },
});

const sliders: ComputedRef<TopicSliderItem | void> = computed(() => {
    if (!topics.value) {
        return;
    }

    const levelOneTopics = topics.value.filter(isTopicFirstDegree);
    const levelTwoTopics = topics.value.filter(isTopicSecondDegree);
    const levelThreeTopics = topics.value.filter(isTopicThirdDegree);

    return {
        levelOne: levelOneTopics.map(createTopicSlider),
        levelTwo: levelTwoTopics.map(createTopicSlider),
        levelThree: levelThreeTopics.map(createTopicSlider),
    };
});

getBranch();
getTopics();
</script>

<template>
    <v-page>
        <v-loader :logs="['useBranch', 'getTopicsByBranch']" />

        <template v-if="branch" #header>
            <v-branch-header>
                {{ branch.name }}

                <template #sub-title>Teilgebiete</template>
            </v-branch-header>
        </template>

        <div v-if="sliders?.levelOne && sliders.levelOne.length && branch" class="v-mb-section">
            <h2 class="v-h2">{{ TopicDegree.one }}</h2>

            <v-topic-box-slider
                :key="branch.id"
                :modifier="'v-mt-box-md v-bg-' + branch.color"
                :branch-img="{
                    src: branch.icon?.url || '',
                    alt: branch.icon?.alt || '',
                }"
                :slider-items="sliders.levelOne"
            >
                {{ themeTitle }}
            </v-topic-box-slider>
        </div>

        <div v-if="sliders?.levelTwo && sliders.levelTwo.length && branch" class="v-mb-section">
            <h2 class="v-h2">{{ TopicDegree.two }}</h2>

            <v-topic-box-slider
                :key="branch.id"
                :modifier="'v-mt-box-md v-bg-' + branch.color"
                :branch-img="{
                    src: branch.icon?.url || '',
                    alt: branch.icon?.alt || '',
                }"
                :slider-items="sliders.levelTwo"
            >
                {{ themeTitle }}
            </v-topic-box-slider>
        </div>

        <div v-if="sliders?.levelThree && sliders.levelThree.length && branch">
            <h2 class="v-h2">{{ TopicDegree.three }}</h2>

            <v-topic-box-slider
                :key="branch.id"
                :modifier="'v-mt-box-md v-bg-' + branch.color"
                :branch-img="{
                    src: branch.icon?.url || '',
                    alt: branch.icon?.alt || '',
                }"
                :slider-items="sliders.levelThree"
            >
                {{ themeTitle }}
            </v-topic-box-slider>
        </div>
    </v-page>
</template>
