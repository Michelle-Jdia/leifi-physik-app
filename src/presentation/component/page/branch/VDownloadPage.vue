<script setup lang="ts">
import type { Collection } from '@/data/type/app/collection';
import type { RefresherCustomEvent } from '@ionic/vue';
import type { Ref } from 'vue';
import { onIonViewWillEnter } from '@ionic/vue';
import { computed, ref, watch } from 'vue';
import { $t } from '@/application/translation';
import { ContentType } from '@/data/type/helper/contentType';
import { useDownloadTopicBranch } from '@/connection/helper/download';
import { createRefresherHandler } from '@/connection/helper/refresher';
import { RouteName } from '@/connection/router/routeName';
import { useArticleState } from '@/state/articleState';
import { findArticleInCollection } from '@/store/collectionStore';
import { useDownloadLive } from '@/store/downloadStore';
import { addDownloadToHistory } from '@/store/historyStore';
import { handleLinksClick } from '@/presentation/helper/handleLink';
import { resizeBrowser } from '@/presentation/helper/swiffy';
import VArticleHeader from '@/layout/header/VArticleHeader.vue';
import VPage from '@/layout/page/VPage.vue';
import VInputButton from '@/molecule/form/VInputButton/VInputButton.vue';
import VAccordion from '@/atom/accordion/VAccordion.vue';
import VButton from '@/atom/button/VButton/VButton.vue';
import VCheckbox from '@/atom/checkbox/VCheckbox/VCheckbox.vue';
import VDownloadContentWrapper from '@/atom/contentWrapper/VDownloadContentWrapper.vue';
import VIcon from '@/atom/icon/VIcon/VIcon.vue';
import VLoader from '@/atom/loader/VLoader/VLoader.vue';
import VModal from '@/atom/modal/VModal/VModal.vue';
import VRefresher from '@/atom/refresher/VRefresher/VRefresher.vue';
import VAnimationScript from '@/atom/script/VAnimationScript/VAnimationScript.vue';
import VBundle from '@/atom/script/VBundle/VBundle.vue';
import VH5p from '@/atom/script/VH5p/VH5p.vue';

interface Props {
    id: string;
}
interface DownloadFile {
    id: string;
    url: string;
    size: string;
}

const props = defineProps<Props>();
const isSwiffyReady: Ref<boolean> = ref(false);
const articleState = useArticleState();
const addToCollectionsModal = ref();
const createCollectionModal = ref();
const rawHtmlLinksContainer = ref();
const tempCollectionsToUpdate: Collection[] = [];
const { data: pageData, getData: getDownloadTopicBranch } = useDownloadTopicBranch(props.id);

const { getData: getDownloadLive } = useDownloadLive({
    params: {
        id: props.id,
    },
});

function openCollectionModal(): void {
    articleState.getCollections();
    tempCollectionsToUpdate.length = 0;
    addToCollectionsModal.value?.$el.present();
}

function toggleArticleInTempList(event: CustomEvent, item: Collection): void {
    if (event.detail.checked === true) {
        tempCollectionsToUpdate.push(item);
    }

    if (event.detail.checked === false) {
        const filtered = tempCollectionsToUpdate.filter(
            (collection) => !(collection.id === item.id && collection.type === item.type),
        );

        tempCollectionsToUpdate.length = 0;
        tempCollectionsToUpdate.push(...filtered);
    }
}

function handleModalConfirmation(): void {
    if (!pageData.value || !pageData.value.download) {
        return;
    }

    articleState.saveArticlesToCollections(pageData.value.download, tempCollectionsToUpdate);
}

function handleCreateCollectionSubmit({ value }: { value: string }): void {
    if (!pageData.value || !pageData.value.download || !value) {
        return;
    }

    articleState.createAndAddArticleToNewCollection(pageData.value.download, value);
}

function openCreateCollectionModal(): void {
    // addToCollectionsModal.value?.$el.dismiss();
    createCollectionModal.value?.$el.present();
}

function setSwiffiyToReady() {
    isSwiffyReady.value = true;
}

function getFileSizeInBites(size: number): string {
    const sizes = ['B', 'KB', 'MB', 'GB'];
    const kb = 1024;
    let i = 0;

    do {
        size = size / kb;
        i++;
    } while (size >= kb);

    return `${size.toFixed(2)} ${sizes[i]}`;
}

const downloadedFiles = computed<DownloadFile[]>(() => {
    if (!pageData.value || !pageData.value.download) {
        return [];
    }

    return pageData.value.download.download_file.map((file) => ({
        id: file.id,
        size: `Größe: ${getFileSizeInBites(file.size)}`,
        url: file.url,
    }));
});

async function handleRefresh(event: RefresherCustomEvent): Promise<void> {
    await createRefresherHandler({
        event: event,
        callback: async () => {
            await getDownloadLive();
            await getDownloadTopicBranch();
        },
    });
}

watch(rawHtmlLinksContainer, () => {
    if (!rawHtmlLinksContainer.value) {
        return;
    }

    handleLinksClick(rawHtmlLinksContainer.value);
});

onIonViewWillEnter(async () => {
    addDownloadToHistory(props.id);

    await getDownloadTopicBranch();

    resizeBrowser();
});
</script>

<template>
    <v-page modifier="v-pt-0 md:v-pt-0">
        <template v-if="pageData">
            <v-bundle />

            <v-h5p />

            <v-animation-script :onload="setSwiffiyToReady" />
        </template>

        <v-loader :logs="['useDownload']" />

        <template #refresher>
            <v-refresher @ion-refresh="handleRefresh" />
        </template>

        <template #header>
            <v-article-header
                :id="id"
                :type="ContentType.DOWNLOAD"
                @on-open-add-to-collection="openCollectionModal"
            >
                <template v-if="pageData && pageData.topic && pageData.branch" #pre-title>
                    <router-link
                        :to="{ name: RouteName.BRANCH_ID, params: { id: pageData.branch.id } }"
                    >
                        {{ pageData.branch.name }}
                    </router-link>
                    /
                    <router-link
                        :to="{ name: RouteName.TOPIC_ID, params: { id: pageData.topic.id } }"
                    >
                        {{ pageData.topic.name }}
                    </router-link>
                </template>

                <template v-if="pageData && pageData.download">
                    {{ pageData.download.title }}
                </template>
            </v-article-header>
        </template>

        <template v-if="pageData">
            <h1 class="v-pt-box-xl v-text-h2">{{ $t.download.headline }}</h1>

            <p class="v-p v-mt-box-lg">
                <span class="v-font-bold">Typ:</span>

                {{ pageData.download.download_type }}
            </p>

            <div
                v-if="isSwiffyReady"
                class="site v-mt-box-xl v-border-t v-border-t-gray-80 v-pt-box-xl"
            >
                <div v-html="pageData.download.download_description" />

                <div
                    v-if="pageData.download.animation_code?.rendered_string"
                    v-html="pageData.download.animation_code.rendered_string"
                />

                <template v-if="pageData.download.animation_code?.attachments">
                    <component
                        :is="attachment.tag"
                        v-for="(attachment, indexAttachment) in pageData.download.animation_code
                            .attachments"
                        :key="`${indexAttachment}-task-question-attachment`"
                        :type="attachment.type"
                    >
                        {{ attachment.content }}
                    </component>
                </template>
            </div>

            <div class="site">
                <template v-if="downloadedFiles.length">
                    <template v-for="file in downloadedFiles" :key="file.id">
                        <p class="v-p v-mt-box-lg">{{ file.size }}</p>

                        <a v-if="file.url" :href="file.url" target="_blank">
                            <v-button class="v-w-full" size="small" :is-rounded="false">{{
                                $t.download.fielDownloadButtonLabel
                            }}</v-button>
                        </a>
                    </template>
                </template>
            </div>

            <v-accordion v-if="pageData.download.download_didactic_info.trim().length">
                <template #label>{{ $t.download.accordionLabel }}:</template>

                <v-download-content-wrapper>
                    <div v-html="pageData.download.download_didactic_info" />
                </v-download-content-wrapper>
            </v-accordion>
        </template>

        <v-modal ref="addToCollectionsModal" with-save @on-confirm="handleModalConfirmation">
            <template #title>{{ $t.collection.addCollection }}</template>

            <v-button modifier="v-w-full" @click="openCreateCollectionModal">
                <v-icon src="icon/add-collection.svg" with-space-right />

                {{ $t.collection.createCollection }}
            </v-button>

            <template v-if="articleState.collections && articleState.collections.length">
                <div
                    v-for="collection in articleState.collections"
                    :key="collection.id"
                    class="v-mt-box-md v-flex v-items-center v-rounded v-border v-border-gray-20 v-bg-white v-p-box"
                >
                    <v-checkbox
                        :disabled="
                            findArticleInCollection(collection, props.id, ContentType.DOWNLOAD)
                        "
                        :checked="
                            findArticleInCollection(collection, props.id, ContentType.DOWNLOAD)
                        "
                        modifier="v-mr-auto v-w-full v-h3 v-font-bold"
                        @ion-change="(event) => toggleArticleInTempList(event, collection)"
                    >
                        {{ collection.private_title }}
                    </v-checkbox>
                </div>
            </template>
        </v-modal>

        <v-modal ref="createCollectionModal">
            <template #title>{{ $t.collection.addCollection }}</template>

            <v-input-button
                icon="icon/add-collection.svg"
                :button-text="$t.collection.create"
                :placeholder="$t.collection.collectionName"
                @on-submit="handleCreateCollectionSubmit"
            />
        </v-modal>
    </v-page>
</template>
