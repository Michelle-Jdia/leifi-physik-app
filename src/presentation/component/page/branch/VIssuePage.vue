<script setup lang="ts">
import '@/presentation/style/static/bootstrap.scss';
import type { ComputedRef } from 'vue';
import type { Task } from '@/data/type/app/task';
import { computed, onMounted } from 'vue';
import { useIssue } from '@/store/issueStore';
import { useTasksByIssue, sortTasksByType } from '@/store/taskStore';
import { setIssueHistory } from '@/store/historyStore';
import VPage from '@/layout/page/VPage.vue';
import VArticleHeader from '@/layout/header/VArticleHeader.vue';
import VLoader from '@/atom/loader/VLoader/VLoader.vue';
import VLinkItem from '@/atom/item/VLinkItem/VLinkItem.vue';
import VButton from '@/atom/button/VButton/VButton.vue';
import { getMathjax } from '@/presentation/helper/mathjax';

interface Props {
    id: string;
}

const props = defineProps<Props>();

const { data: issue, getData: getIssue } = useIssue({
    params: {
        id: props.id,
    },
});

const { data: tasks, getData: getTasksByIssue } = useTasksByIssue({
    params: {
        issueId: props.id,
    },
});

const tasksByType: ComputedRef<Record<string, Task[]> | void> = computed(() => {
    if (!tasks.value) {
        return;
    }

    return sortTasksByType(tasks.value);
});

onMounted(getMathjax);

setIssueHistory(props.id);
getTasksByIssue();
getIssue();
</script>

<template>
    <v-page>
        <v-loader :logs="['useIssue', 'useTasksByIssue']" />

        <template v-if="issue" #header>
            <v-article-header :id="props.id" type="issue" modifier="v-py-box-md">
                {{ issue.title }}
            </v-article-header>
        </template>

        <div v-if="issue">
            <h1 class="v-text-h2">{{ issue.issue_category }}</h1>

            <h2 class="v-mt-box-sm v-text-h4">{{ issue.issue_teaser_title }}</h2>

            <p class="v-mt-box-xl" v-html="issue.issue_teaser_content" />

            <hr class="v-mt-[3.3rem]" />

            <div class="v-mt-[1.8rem]">
                <div
                    v-for="(paragraph, index) in issue.issue_paragraphs"
                    :key="`${index}-issue-paragraph`"
                    v-html="paragraph"
                />
            </div>

            <div class="v-mt-box-md">
                <div
                    v-for="(paragraph_ext, index) in issue.issue_paragraphs_ext"
                    :key="`${index}-issue-paragraph_ext`"
                    v-html="paragraph_ext"
                />
            </div>
        </div>

        <div v-if="tasksByType" class="v-mt-section-md">
            <h2 class="v-text-h2">Aufgaben</h2>

            <div v-for="(sortedTasks, type) in tasksByType" :key="type">
                <h4 class="v-mt-box-xl v-text-h4">{{ type }}</h4>

                <div v-for="task in sortedTasks" :key="task.id">
                    <router-link :to="{ name: 'task-id', params: { id: task.id } }">
                        <v-link-item class="v-mt-[.5rem]">{{ task.title }}</v-link-item>
                    </router-link>
                </div>
            </div>
        </div>

        <v-button modifier="v-w-[12rem]">Vorheriger Artikel</v-button>

        <v-button modifier=" v-w-[12rem]" class="v-text">NÃ¤chster Artikel</v-button>
    </v-page>
</template>
