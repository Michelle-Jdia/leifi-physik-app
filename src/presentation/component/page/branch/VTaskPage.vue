<script setup lang="ts">
import { onMounted } from 'vue';
import { onIonViewWillEnter } from '@ionic/vue';
import type { Issue } from '@/data/type/app/issue';
import { createReactiveData } from '@/connection/helper/fetcher';
import { useTask } from '@/store/taskStore';
import { useIssue } from '@/store/issueStore';
import { getMathjax } from '@/presentation/helper/mathjax';
import VArticleHeader from '@/layout/header/VArticleHeader.vue';
import VPage from '@/layout/page/VPage.vue';
import VLoader from '@/atom/loader/VLoader/VLoader.vue';
import VLinkItem from '@/atom/item/VLinkItem/VLinkItem.vue';
import VAccordion from '@/atom/accordion/VAccordion.vue';

interface Props {
    id: string;
}
const props = defineProps<Props>();

const { data: task, getData: getTask } = useTask({
    params: {
        id: props.id,
    },
});

const { data: issueData, getData: getIssueData } = createReactiveData<never, Issue | void>(
    async () => {
        await getTask();

        if (!task.value) {
            return;
        }

        const { data: issue, getData: getIssue } = useIssue({
            params: {
                id: task.value.referenced_issue[0], // @TODO: array of related issues?
            },
        });

        await getIssue();

        if (!issue.value) {
            return;
        }

        return issue.value;
    },
)();

getTask();
onIonViewWillEnter(getIssueData);
onMounted(getMathjax);
</script>

<template>
    <v-page>
        <v-loader :logs="['useTask']" />

        <template v-if="task" #header>
            <v-article-header :id="task.id" type="task">{{ task.title }}</v-article-header>
        </template>

        <div v-if="task">
            <h3 class="v-h3 v-my-box v-text-primary">{{ task.level }}</h3>

            <div
                v-for="(question, index) in task.question"
                :key="`${index}-task-question`"
                class="question"
                v-html="question"
            />

            <v-accordion modifier="v-my-box">
                <template #accordion-label>
                    <h3 class="!v-text-h3 v-text-primary v-underline">Answers | LÃ¶sung</h3>
                </template>

                <div
                    v-for="(answer, index) in task.answer"
                    :key="`${index}-task-answer`"
                    v-html="answer"
                />
            </v-accordion>

            <h3 class="v-h3 v-mt-box v-text-primary">Paragraph [H5P]</h3>

            <div class="v-mt-box" v-html="task.paragraphs" />

            <template v-if="issueData">
                <h3 class="v-h3 v-mt-box v-text-primary">Grundwissen</h3>

                <router-link :to="{ name: 'issue-id', params: { id: issueData.id } }">
                    <v-link-item class="v-mt-box">{{ issueData.title }}</v-link-item>
                </router-link>
            </template>
        </div>
    </v-page>
</template>
