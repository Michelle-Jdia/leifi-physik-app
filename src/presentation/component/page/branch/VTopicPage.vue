<script setup lang="ts">
import type { Branch } from '@/data/type/app/branch';
import type { Issue } from '@/data/type/app/issue';
import { onIonViewWillEnter } from '@ionic/vue';
import { createReactiveData } from '@/connection/helper/fetcher';
import {
    useIssuesByTopic,
    isIssueFirstCategory,
    isIssueSecondCategory,
    isIssueThirdCategory,
    isIssueFourthCategory,
    IssueCategory,
} from '@/store/issueStore';
import { setFavoriteIssue, useFavoriteIssues } from '@/store/favoriteStore';
// import { getTasksByTopic } from '@/store/taskStore';
import { useBranch } from '@/store/branchStore';
import { useTopic } from '@/store/topicStore';
import VPage from '@/layout/page/VPage.vue';
import VBranchHeader from '@/layout/header/VBranchHeader/VBranchHeader.vue';
import VArticleInfoSlide from '@/molecule/slide/VArticleInfoSlide/VArticleInfoSlide.vue';
import VLoader from '@/atom/loader/VLoader/VLoader.vue';

interface Props {
    id: string;
}

interface PageData {
    branch: Branch | void;
    categories: {
        title: string;
        items: Issue[];
    }[];
}

const props = defineProps<Props>();

function sortIssuesByWeight(issue1: Issue, issue2: Issue) {
    return issue1.issue_weight - issue2.issue_weight;
}

const { data: topic, getData: getTopic } = useTopic({
    params: {
        id: props.id,
    },
});

const { data: issues, getData: getIssue } = useIssuesByTopic({
    params: {
        topicId: props.id,
    },
});

const { data: favorites, getData: getFavorites } = useFavoriteIssues();

// // const { data: tasks, getData: getTasksByTopicData } = getTasksByTopic({
// //     params: {
// //         topicId: props.id,
// //     },
// // });

// getTasksByTopic();

const { data: pageData, getData: getPageData } = createReactiveData<
    never,
    Partial<PageData> | void
>(async () => {
    await getTopic();

    if (!topic.value) {
        return;
    }

    const { data: branch, getData: getBranch } = useBranch({
        params: {
            id: topic.value.referenced_branch,
        },
    });

    await Promise.all([getIssue(), getBranch()]);

    if (!branch.value && !issues.value) {
        return;
    }

    if (!issues.value) {
        return {
            branch: branch.value,
        };
    }

    const one = issues.value.filter(isIssueFirstCategory).sort(sortIssuesByWeight);
    const two = issues.value.filter(isIssueSecondCategory);
    const three = issues.value.filter(isIssueThirdCategory);
    const four = issues.value.filter(isIssueFourthCategory);

    const categories = [
        {
            title: `${IssueCategory.one} & Aufgaben`,
            items: one,
        },
        {
            title: IssueCategory.two,
            items: two,
        },
        {
            title: IssueCategory.three,
            items: three,
        },
        {
            title: IssueCategory.four,
            items: four,
        },
    ];

    if (!branch.value) {
        return {
            categories,
        };
    }

    return {
        branch: branch.value,
        categories,
    };
})();

async function updateFavorite(id: string, closeSlide: () => void): Promise<void> {
    await setFavoriteIssue(id);
    getFavorites();
    closeSlide();
}

onIonViewWillEnter(getPageData);
onIonViewWillEnter(getFavorites);
</script>

<template>
    <v-page modifier="v-px-0 md:v-px-0">
        <v-loader :logs="['useTopic', 'useTasksByTopic', 'useIssuesByTopic', 'useBranch']" />

        <template #header>
            <v-branch-header>
                <template v-if="topic">
                    {{ topic.name }}
                </template>
            </v-branch-header>
        </template>

        <template v-if="pageData && pageData.categories">
            <div v-for="(slider, index) in pageData.categories" :key="slider.title">
                <h2
                    v-if="slider.items.length"
                    :class="`v-h3 v-border-b v-border-b-gray-50 v-px-box v-font-bold ${
                        index !== 0 ? 'v-py-box-xl' : 'v-pb-box-xl'
                    }`"
                >
                    {{ slider.title }}
                </h2>

                <div
                    v-for="issue in slider.items"
                    :key="issue.id"
                    class="v-border-b v-border-b-gray-50"
                >
                    <router-link
                        v-if="pageData.branch"
                        :to="{ name: 'issue-id', params: { id: issue.id } }"
                    >
                        <v-article-info-slide
                            :id="issue.id"
                            modifier="border-y"
                            :title="issue.title"
                            :sub-title="issue.issue_category"
                            :is-favorite="favorites?.includes(issue.id)"
                            :branch="{
                                title: pageData.branch.name,
                                color: `v-bg-${pageData.branch.color}`,
                            }"
                            @on-favorite="updateFavorite"
                        />
                    </router-link>
                </div>
            </div>
        </template>
    </v-page>
</template>
