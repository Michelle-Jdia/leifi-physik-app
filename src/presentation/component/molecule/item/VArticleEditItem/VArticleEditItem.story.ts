import type { Meta, StoryFn } from '@storybook/vue3';
import { branch } from '@/presentation/static/branch';
import { topic } from '@/presentation/static/topic';
import VArticleEditItem from '@/molecule/item/VArticleEditItem/VArticleEditItem.vue';

export default {
    component: VArticleEditItem,
} as Meta<typeof VArticleEditItem>;

const Template: StoryFn<typeof VArticleEditItem> = (args) => ({
    components: {
        VArticleEditItem,
    },
    setup() {
        return {
            args,
        };
    },
    template: `<v-article-edit-item v-bind="args"></v-article-edit-item>`,
});

export const Base = Template.bind({});

export const WithCopy = Template.bind({});

const defaultArgs = {
    id: '1',
    title: branch.title,
    subtitle: topic.subtitle,
};

Base.args = defaultArgs;

WithCopy.args = {
    ...defaultArgs,
    withCopy: true,
};
