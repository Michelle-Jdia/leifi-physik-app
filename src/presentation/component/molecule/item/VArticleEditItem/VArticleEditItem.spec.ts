import { render } from '@testing-library/vue';
import { describe, expect, test } from 'vitest';
import { testSnapshot } from '@/presentation/helper/test';
import { branch } from '@/presentation/static/branch';
import { topic } from '@/presentation/static/topic';
import VArticleEditItem from '@/molecule/item/VArticleEditItem/VArticleEditItem.vue';

describe('VArticleEditItem', () => {
    const options = {
        props: {
            id: '1',
            title: branch.title,
            subtitle: topic.subtitle,
        },
        global: {
            stubs: {
                ionIcon: true,
            },
        },
    };

    test('shows the correct text', () => {
        const { getByText, queryByTestId } = render(VArticleEditItem, options);

        getByText(branch.title);
        getByText(topic.subtitle);

        expect(queryByTestId('article-edit-item-copy')).toBeFalsy();
    });

    test('doesnt show the text if it was not passed', () => {
        const { queryByText, getByTestId } = render(VArticleEditItem, {
            ...options,
            props: {
                ...options.props,
                title: '',
                subtitle: '',
                withCopy: true,
            },
        });

        expect(queryByText(branch.title)).toBeFalsy();
        expect(queryByText(topic.subtitle)).toBeFalsy();

        getByTestId('article-edit-item-copy');
    });

    testSnapshot(VArticleEditItem, options);
});
