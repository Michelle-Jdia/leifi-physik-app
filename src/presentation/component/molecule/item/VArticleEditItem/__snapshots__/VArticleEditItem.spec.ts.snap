// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VArticleEditItem > snapshot 1`] = `
<div
  class="article-edit-item v-flex v-items-center v-rounded v-bg-white v-p-box"
  lines="none"
>
  <v-checkbox-stub
    modifier="v-mr-auto v-w-full"
    value="1"
  >
    Mechanik 
    <div
      class="v-h5 v-text-ty-helper"
    >
      Grundwissen
    </div>
  </v-checkbox-stub>
  <div
    class="v-ml-box v-flex"
  >
    <!--v-if-->
    <v-icon-stub
      data-testid="article-edit-item-close"
      modifier="v-cursor-pointer"
      size="default"
      src="icon/close.svg"
      withspaceleft="false"
      withspaceright="false"
    />
    <anonymous-stub
      class="v-ml-box-xl"
      routerlink="Symbol()"
      slot="end"
    >
      <v-icon-stub
        data-testid="article-edit-item-reorder"
        modifier="false"
        size="default"
        src="icon/dots.svg"
        withspaceleft="false"
        withspaceright="false"
      />
    </anonymous-stub>
  </div>
</div>
`;
